extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -112,7 +111,24 @@ public QueryDefinition makeQueryDefinition(
         false
     );
 
-    dataSourcePlan.getSubQueryDefBuilder().ifPresent(queryDefBuilder::addAll);
+    final QueryDefinitionBuilder builder = QueryDefinition.builder(queryId);
+    dataSourcePlan.getSubQueryDefBuilder().ifPresent(builder::addAll);
+
+    QueryDefinitionBuilder queryDefBuilder = QueryDefinition.builder(queryId);
+    if (isEmptyOverPresent) {
+      // If window has an empty over, we want a single worker to process entire data for window function evaluation.
+      // To achieve that, we are overriding the shuffle spec of the last stage to MixShuffleSpec.
+      int previousStageNumber = dataSourcePlan.getSubQueryDefBuilder().get().build().getFinalStageDefinition().getStageNumber();
+      for (final StageDefinition stageDef : dataSourcePlan.getSubQueryDefBuilder().get().build().getStageDefinitions()) {
+        if (stageDef.getStageNumber() == previousStageNumber) {
+          queryDefBuilder.add(StageDefinition.builder(stageDef).shuffleSpec(MixShuffleSpec.instance()));
+        } else {
+          queryDefBuilder.add(StageDefinition.builder(stageDef));
+        }
+      }
+    } else {
+      queryDefBuilder = builder;
+    }
```

Comment:
it would be nice to move this into a separate method so that the name of the method could tell the brief story; and the comment could live as an apidoc to it

Reply:
Makes sense, have made the change.